### ISSUE TITLE
Error inferring the package name of a class that results in compilation failure

### ISSUE THREAD

**Issue by RaulGracia** [greenjoe/lambdaFromString#8]
Hi all,

I have found a problem using this library when compiling the following lambda:
`....map(word -> new SimpleEntry<String, Long>(word, (long) 1))...`

The problem consists of an undesired repetition in the package name of the SimpleEntry class, which makes the compilation process to fail (i.e., java.util.AbstractMap.java.util.AbstractMap$SimpleEntry does not exist). Also, instead of a '.', the SimpleEntry name was preceded of a '$' symbol, which is also a source of problems.

After some debugging, I found where the problem arises: the createLambda method in the LambdaFactory class. The problem comes when the typeReference.toString() is invoked, returning a malformed class name for SimpleEntry class.

To bypass this problem at the moment, I created a "sanity check" method that inspects if the class name returned by typeReference.toString() has duplicates or not, and deletes them in the affirmative case. This works for my case as well as for the test cases provided in the library. However, maybe this can be solved in a much better way by addressing the root cause of why typeReference.toString() sometimes does not behave properly.

Thanks a lot,
Raúl.

**Comment by greenjoe** [comment]
Thank you for your report, I reproduced the behavior in a test case. As you said it happens because `typeReference.toString()` just invokes `java.lang.reflect.Type,toString()` and it doesn't work well in case of SimpleEntry. It looks like using https://commons.apache.org/proper/commons-lang/apidocs/src-html/org/apache/commons/lang3/reflect/TypeUtils.html#line.1661 from apache commons lang3 solves the issue, but I will see whether I can avoid that additional dependency.

**Comment by greenjoe** [comment]
Well it looks like I will have to add the apache commons dependency, I should release a new version of the library soon :) In the meantime the best workaround seems to provide a type representation in a String using `DynamicTypeReference` (which unfortunately makes the call quite verbose as the type is specified three times) 
```
BiFunction<String, Long, SimpleEntry<String, Long>> lambda = ( BiFunction<String, Long, SimpleEntry<String, Long>>) factory.createLambdaUnchecked(
    code, new DynamicTypeReference("BiFunction<String, Long, SimpleEntry<String, Long>>"));
```

**Comment by RaulGracia** [comment]
Hi @greenjoe,

I assume that with the addition of the new library the way of using "factory.createLambdaUnchecked" will remain the same as before, right? In the affirmative case, I will just wait to the update from your side as my code already relies on the previous way of compiling lambdas and the fix I did does the job meanwhile :)

Thanks a lot,
Raúl.

**Comment by RaulGracia** [comment]
Hi @greenjoe,

Could you tell me the status of adding the the apache commons library in the project? Is just to know if I should continue working with the modified version of the library or if I can work with the master branch and close the issue.

Thanks!
Raúl.

**Comment by greenjoe** [comment]
It turned out that Apache Commons Lang3 had a bug in the method I wanted to use to describe types. The bug is already fixed in their master branch by commit https://github.com/apache/commons-lang/commit/1661e5519c4836a5a940b13b7797263443156fc9 but the fix is not yet present in the latest release which is 3.5. 
I looked at their mailing list and for a while it looked like a new release is coming soon http://mail-archives.apache.org/mod_mbox/commons-dev/201703.mbox/browser so I thought I might wait for that fix, but as it doesn't happen I just used 3.5 and included some workaround for cases with array types. I will change that when apache commons are released.

I am sorry it took so long, the issue should be fixed in 1.5 that I've just released into maven central (it can take a few hours until it is visible). Thanks again for report and I hope the new version will fork fine for you, let me know if there are still some issues, best regards :-)
